### PR TITLE
feat: publish org file as a message to a group

### PR DESCRIPTION
**Pull request by lujun9972**

### CODE REVIEW

File: habitica.el [hunk 1]
```diff
@@ -343,6 +349,31 @@ Options are 0.1, 1, 1.5, 2; eqivalent of Trivial, Easy, Medium, Hard."
...
+                           (string= name (funcall get-group-name-fn group))))
+         (the-group (cl-find-if the-group-p-fn groups)))
+    (setq xxx the-group)
+    (cdr (assoc-string "_id" the-group))))
+
+(defun habitica-api-chat2group (message group-id)
+  "Post chat MESSAGE to a group."
+  (habitica--send-request (format "/groups/%s/chat" group-id) "POST" (concat "message=" (url-encode-url message))))
+
+;; (habitica-api-chat2group "published from Emacs" "4e1b6290-3026-4b52-9c55-8c58f4b5c54d")
```
Comment (abrochard) [inline habitica.el]: probably best to remove the test comment

File: habitica.el [hunk 2]
```diff
@@ -343,6 +349,31 @@ Options are 0.1, 1, 1.5, 2; eqivalent of Trivial, Easy, Medium, Hard."
...
+(defun habitica-api-get-group-id (&optional name)
+  "Get the id of group which name is NAME"
+  (let* ((groups (habitica-api-get-groups))
+         (get-group-name-fn (lambda (group)
+                              (cdr (assoc-string "name" group))))
+         (name (or name (completing-read "Select the group: "  (mapcar get-group-name-fn groups))))
+         (the-group-p-fn (lambda (group)
+                           (string= name (funcall get-group-name-fn group))))
+         (the-group (cl-find-if the-group-p-fn groups)))
+    (setq xxx the-group)
```
Comment (abrochard) [inline habitica.el]: is that for debug?

File: habitica.el [hunk 3]
```diff
@@ -343,6 +349,31 @@ Options are 0.1, 1, 1.5, 2; eqivalent of Trivial, Easy, Medium, Hard."
...
   (habitica--send-request (format "/user/feed/%s/%s" pet food) "POST" ""))
 
+(defun habitica-api-get-groups (&rest types)
+  "Get groups which type is in TYPES."
+  (let* ((types (or types '("party" "guilds" "privateGuilds" "publicGuilds" "tavern")))
+         (types (string-join types ",")))
+    (message "types=%s" types)
+    (habitica--send-request (format "/groups?type=%s" (url-encode-url types)) "GET" "")))
+
+(defun habitica-api-get-group-id (&optional name)
```
Comment (abrochard) [inline habitica.el:358]: what happens if the name is empty?

File: habitica.el [hunk 4]
```diff
@@ -343,6 +349,31 @@ Options are 0.1, 1, 1.5, 2; eqivalent of Trivial, Easy, Medium, Hard."
...
+  "Get the id of group which name is NAME"
+  (let* ((groups (habitica-api-get-groups))
```
Comment (abrochard) [inline habitica.el:360]: might be worth caching those results if they aren't going to change frequently. but totally optional and can be a follow up.

File: habitica.el (hunk 3 above)
Comment (lujun9972) [reply habitica.el:358]:
```emacs-lisp
         (name (or name (completing-read "Select the group: "  (mapcar get-group-name-fn groups))))
```
if `name` is empty,  the function will list out all the group names and ask the user to select a group.

File: habitica.el (hunk 4 above)
Comment (lujun9972) [reply habitica.el:360]: The result will be changing as you join or leave a group.

**Review by abrochard — APPROVED** [review]
Thank you!

**Pull request merged**